nemo/collections/asr/modules/wav2vec_modules.py

```diff
@@ -17,290 +17,312 @@
+        for conv in self.conv_layers:
+            x = conv(x)
+        return x
+
+    def normalize(self, source, lengths):
+        with torch.no_grad():  # Normalizes audio source
+            for i in range(lengths.size(0)):
+                orig = source[i, : lengths[i]]
+                norm = F.layer_norm(orig, orig.shape)
+                source[i, : lengths[i]] = norm
+        return source
+
+    def forward(self, input_signal, length):
+        if self.normalize_input:
+            input_signal = self.normalize(input_signal, length)
+
+        # BxT -> BxCxT
+        processed_signal = input_signal.unsqueeze(1)
+
+        # Applies grad mult scaling
+        if self.grad_mult > 0:
+            processed_signal = self.apply_layers(processed_signal)
+            if self.grad_mult != 1.0:
+                processed_signal = GradMultiply.apply(processed_signal, self.grad_mult)
         else:
-            self.weight_proj = nn.Linear(self.input_dim, groups * num_vars)
-            nn.init.normal_(self.weight_proj.weight, mean=0, std=1)
-            nn.init.zeros_(self.weight_proj.bias)
+            with torch.no_grad():  # 0 indicates frozen feature encoder
+                processed_signal = self.apply_layers(processed_signal)
+
+        processed_signal = processed_signal.transpose(1, 2)  # B,T,C
+        # Project to embedding
+        if self.post_extract_proj is not None:
+            processed_signal = self.post_extract_proj(processed_signal)
+
+        # Adding normalization for output
+        if self.mode == "layer_norm":
+            processed_signal = self.layer_norm(processed_signal)
+
+        processed_signal = processed_signal.transpose(1, 2)  # B,C,T
+
+        # Feature lengths will have been changed through convolutions
+        processed_signal_length = self.get_lengths(audio_lengths=length)
 
-        assert len(temp) == 3, "Quantize temperature should be a tuple of 3 elements: (start, stop, decay factor)"
+        return processed_signal, processed_signal_length
 
-        self.max_temp, self.min_temp, self.temp_decay = temp
-        self.curr_temp = self.max_temp
-        self.codebook_indices = None
+    def get_lengths(self, audio_lengths):
+        # converts audio lengths to timestep lengths
+        for conv in self.layer_cfg:
+            kernel = conv[1]
+            stride = conv[2]
+            audio_lengths = (
+                torch.div(audio_lengths - kernel, stride, rounding_mode='floor') + 1
+            )  # from pytorch documentation
+        return audio_lengths
 
-    def set_num_updates(self, num_updates):
-        self.curr_temp = max(self.max_temp * self.temp_decay ** num_updates, self.min_temp)
 
-    def get_codebook_indices(self):
-        if self.codebook_indices is None:
-            from itertools import product
+class Wav2VecTransformerEncoder(TransformerEncoder):
+    def __init__(self, pos_embed: DictConfig, transformer: DictConfig, layer_drop: float = 0.0):
```

Comment:
Document what are valid values for `transformer` dictconfig here. Similar to jasper block